### PR TITLE
Replace UDT feature flag with cyclic UDT feature flag [HZ-3496]

### PR DESCRIPTION
**Pull request by Fly-Style**
User-defined types are recognized as production-ready for 5.4, except cyclic UDTs, which are still in beta. 
This patch replaces UDT feature flag with the **cyclic** UDT feature flag, which prevents querying of such data, but doesn't forbid the creation of such type. If the user tries to query source with cyclic UDT, an exception with a suggestion to enable them via feature-flag will be thrown.

### CODE REVIEW

**Comment by Fly-Style** [comment]
run-lab-run

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/OptimizerContext.java [hunk 1]
```diff
@@ -103,31 +103,33 @@ private OptimizerContext(
...
             List<List<String>> schemaPaths,
             List<Object> arguments,
             IMapResolver iMapResolver,
-            SqlSecurityContext securityContext
+            SqlSecurityContext securityContext,
+            boolean cyclicUserTypesAreAllowed
     ) {
         Prepare.CatalogReader catalogReader = createCatalogReader(rootSchema, schemaPaths);
-        HazelcastSqlValidator validator = new HazelcastSqlValidator(catalogReader, arguments, iMapResolver);
+        HazelcastSqlValidator validator = new HazelcastSqlValidator(
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/OptimizerContext.java:131]: will conflict with HZ-3554

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/OptimizerContext.java [hunk 2]
```diff
@@ -103,31 +103,33 @@ private OptimizerContext(
...
      */
     public static OptimizerContext create(
             SqlCatalog schema,
             List<List<String>> searchPaths,
             List<Object> arguments,
             IMapResolver iMapResolver,
-            SqlSecurityContext securityContext
-    ) {
+            SqlSecurityContext securityContext,
+            boolean cyclicUserTypesAreAllowed) {
```
Comment (shultseva) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/OptimizerContext.java:115]: format

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/BasicNestedFieldsTest.java [hunk 3]
```diff
@@ -76,7 +76,7 @@ public static Object[] parameters() {
     @BeforeClass
     public static void beforeClass() {
         Config config = smallInstanceConfig()
-                .setProperty(SQL_CUSTOM_TYPES_ENABLED.getName(), "true");
+                .setProperty(ClusterProperty.SQL_CUSTOM_CYCLIC_TYPES_ENABLED.getName(), "true");
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/BasicNestedFieldsTest.java:79]: we should test first and foremost behavior in default config. I do not want to end up with situation when normal UDT usage (tested in this class) requires the flag. The flag should be enabled only in tests that really need it.

Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/BasicNestedFieldsTest.java:79]: `test_circularlyRecurrentTypes` uses cyclic UDT. I'll move it.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java [hunk 4]
```diff
@@ -347,10 +358,18 @@ public Void visit(SqlCall call) {
...
-                .getRowType();
+
+    private void validateUpsertRowType(@Nullable SqlIdentifier table) {
+        if (doesRowTypeContainCyclicTypes(table)) {
+            throw QueryException.error("Upserts are not supported for cyclic data type columns");
+        }
+    }
+
+    private boolean doesRowTypeContainCyclicTypes(@Nullable SqlIdentifier table) {
+        if (table == null) {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java]: in what case `table` can be null?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/BasicNestedFieldsTest.java (hunk 3 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/BasicNestedFieldsTest.java]: so you can remove the property

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/CyclicUDTAreNotAllowedTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static com.hazelcast.jet.sql.impl.type.BasicNestedFieldsTest.createJavaMapping;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class CyclicUDTAreNotAllowedTest extends SqlTestSupport {
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        initializeWithClient(1, null, null);
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/CyclicUDTAreNotAllowedTest.java]: you don't use client in these tests

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/UdtObjectToJsonFunctionTest.java [hunk 6]
```diff
@@ -46,7 +46,7 @@ public class UdtObjectToJsonFunctionTest extends SqlJsonTestSupport {
     @BeforeClass
     public static void beforeClass() {
         Config config = smallInstanceConfig()
-                .setProperty(SQL_CUSTOM_TYPES_ENABLED.getName(), "true");
+                .setProperty(ClusterProperty.SQL_CUSTOM_CYCLIC_TYPES_ENABLED.getName(), "true");
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/UdtObjectToJsonFunctionTest.java:49]: flag not needed by most (all?) tests

**Review by k-jamroz — APPROVED** [review]
LGTM, however you need to remove the flag from some tests that should not need it (and maybe move some tests).
In principle, I would prefer entire validation done in `HazelcastSqlValidator`, not partially in optimization rule. Validator is a more natural place for that. But using rule is acceptable and might be more straightforward. Also this validation hopefully will not be needed sometime in the future.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/UdtObjectToJsonFunctionTest.java (hunk 6 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/UdtObjectToJsonFunctionTest.java:49]: It is needed only for one test, but it makes sense to leave it as it is: this test is a unit test for corresponded function.

**Comment by shultseva** [comment]
Consider moving checks in HazelcastSqlValidator in the future. Possible solution to receive all tables used in sql request:
`
var tablesId = super.namespaces.values().stream()
                    .filter(namespace -> namespace.isWrapperFor(IdentifierNamespace.class))
                    .map(namespace -> namespace.unwrap(IdentifierNamespace.class).getId())
                    .collect(Collectors.toList());
`

**Review by shultseva — APPROVED** [review]
Consider moving the validations to HazelcastSqlValidator in the future to keep the validation logic in one place. Possible solution to receive all tables used in sql request:
`var tablesId = super.namespaces.values().stream() .filter(namespace -> namespace.isWrapperFor(IdentifierNamespace.class)) .map(namespace -> namespace.unwrap(IdentifierNamespace.class).getId()) .collect(Collectors.toList());`

**Pull request merged**